### DISCUSSION TITLE
WiX v4 equivalent of v3 Condition Element

### DISCUSSION THREAD

**Discussion by dmcmenamin-kt** [wixtoolset/issues#7483] [Questions] [answered]
I've been reading along in the WiX tutorial (based on WiX v3) to help me create a Wix v4 project for my application. I would like my installer to be able to check that the appropriate version of the .NET Framework is installed on the target, as described in the tutorial [here](https://www.firegiant.com/wix/tutorial/net-and-net/framed-by-.net/).

In the tutorial, they make use of a [Condition](https://wixtoolset.org/docs/v3/xsd/wix/condition/) element within a Product element:
```
<?xml version='1.0' encoding='windows-1252'?>
<Wix xmlns='http://schemas.microsoft.com/wix/2006/wi'>
  <Product Name='Foobar 1.0' Id='YOURGUID-86C7-4D14-AEC0-86416A69ABDE' UpgradeCode='YOURGUID-7349-453F-94F6-BCB5110BA4FD'
    Language='1033' Codepage='1252' Version='1.0.0' Manufacturer='Acme Ltd.'>
    ...
    <PropertyRef Id="NETFRAMEWORK20_SP_LEVEL"/>
    <Condition Message='This setup requires the .NET Framework 2.0 with Service Pack 1 installed.'>
      <![CDATA[Installed OR (NETFRAMEWORK20_SP_LEVEL AND NETFRAMEWORK20_SP_LEVEL = "#1")]]>
    </Condition>
    ...
  </Product>
</Wix>
```

In WiX v4, it seems the Condition element is no longer a valid element in the Wxs schema. I've found a few bits of information, but I'm having trouble piecing them together into how I would replicate this behavior in WiX v4:

* There is now a [Condition](https://wixtoolset.org/docs/schema/bal/condition/) element in the Bal schema. Is this intended to replace the functionality of the v3 Condition element? If so, is it now limited to being a child of just Bundle and Fragment elements instead of the previous Component, Control, Feature, Fragment, PermissionEx, and Product elements? I am currently not using a Bundle, so would I just have to move this element to be within a Fragment instead of a ~~Product~~Package?
* I've seen a few [posts](https://github.com/orgs/wixtoolset/discussions/7197) about moving the condition being evaluated inside a Condition attribute of the Condition element, but am uncertain about the exact syntax modifications. Would this be the equivalent of the condition in the above code? I get a `WIX1150: The binder doesn't know how to place the following symbol into the output: SymbolName: 'WixBalCondition', Id: '` warning when trying to use the condition in my project. Also, is the single `=` in the expression correct instead of `==`?
```
<bal:Condition Message='This setup requires the .NET Framework 2.0 with Service Pack 1 installed.' Condition='Installed OR (NETFRAMEWORK20_SP_LEVEL AND NETFRAMEWORK20_SP_LEVEL = &quot;#1&quot;)' />
```
* Is simply putting a Condition within a Fragment sufficient to prevent installation of the package? I have tried to create an expression that should evaluate to false, e.g. `#1 == #2`, but installing my MSI proceeds without issue.
* There is now a [DotNetCompatibilityCheck](https://wixtoolset.org/docs/schema/netfx/dotnetcompatibilitycheck/) element in the Bal schema. Is this intended to replace using the `NETFRAMEWORK...` properties? It looks like it could be used to check ASP.NET, .NET Core, and .NET Desktop runtimes but not .NET Framework?

Thanks in advance for any help, I'm very much new to WiX.

**Comment by barnson** [comment, marked as answer]
You want https://wixtoolset.org/docs/schema/wxs/launch/.

**Reply by dmcmenamin-kt** [reply]
Thanks, Bob! This was just what I needed.

For future readers, this is what I ended up putting within my Package element to get the behavior I wanted (except checking for .NET Framework 4.8 instead of .NET Framework 2.0 service pack level):
```
<PropertyRef Id="WIX_IS_NETFRAMEWORK_48_OR_LATER_INSTALLED"/>
<Launch Condition="(WIX_IS_NETFRAMEWORK_48_OR_LATER_INSTALLED = 1)" Message=".NET Framework 4.8 must be installed prior to installing this application" />
```
